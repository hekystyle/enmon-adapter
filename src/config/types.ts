/* eslint-disable max-classes-per-file */
import { Type } from 'class-transformer';
import { IsDefined, IsEnum, IsMongoId, IsNotEmpty, IsString, IsUrl, ValidateNested } from 'class-validator';
import { EnmonEnv } from '../enmon/ApiClient.js';

class ConfigEnmon {
  @IsDefined()
  @IsEnum(EnmonEnv)
  readonly env!: EnmonEnv;

  @IsDefined()
  @IsString()
  @IsMongoId()
  readonly customerId!: string;

  @IsDefined()
  @IsString()
  @IsNotEmpty()
  readonly devEUI!: string;

  @IsDefined()
  @IsString()
  @IsNotEmpty()
  readonly token!: string;
}

class ConfigThermometerV1 {
  @IsDefined()
  @IsString()
  @IsUrl()
  readonly dataSourceUrl!: string;

  @IsDefined()
  @ValidateNested()
  @Type(() => ConfigEnmon)
  readonly enmon!: Readonly<ConfigEnmon>;
}

export class ConfigThermometer extends ConfigThermometerV1 {
  @IsDefined()
  @IsString()
  @IsEnum(['UNI7xxx'])
  readonly model!: 'UNI7xxx';
}

class ConfigWattrouter {
  @IsDefined()
  @IsString()
  @IsUrl()
  readonly baseURL!: string;

  @IsDefined()
  @ValidateNested()
  @Type(() => ConfigEnmon)
  readonly enmon!: Readonly<ConfigEnmon>;
}

export class Config {
  @IsDefined()
  @ValidateNested()
  @Type(() => ConfigThermometerV1)
  readonly thermometer!: Readonly<ConfigThermometerV1>;

  @IsDefined()
  @ValidateNested({ each: true })
  @Type(() => ConfigThermometer)
  readonly thermometers!: readonly ConfigThermometer[];

  @IsDefined()
  @ValidateNested()
  @Type(() => ConfigWattrouter)
  readonly wattrouter!: Readonly<ConfigWattrouter>;
}
