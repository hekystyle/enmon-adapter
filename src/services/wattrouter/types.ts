/* eslint-disable max-classes-per-file */
import { Transform, Type } from 'class-transformer';
import { IsDateString, IsNumber, IsDefined, ValidateNested } from 'class-validator';

export class AllTimeStats {
  /** Counted from this date in format YYYY-MM-DD */
  @IsDefined()
  @IsDateString()
  SAD!: string;

  /** Surplus on Phase L1 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAS1!: number;

  /** Consumption in high tariff on Phase L1 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAH1!: number;

  /** Consumption in low tariff on Phase L1 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAL1!: number;

  /** Production on Phase L1 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAP1!: number;

  /** Surplus on Phase L2 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAS2!: number;

  /** Consumption in high tariff on Phase L2 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAH2!: number;

  /** Consumption in low tariff on Phase L2 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAL2!: number;

  /** Production on Phase L2 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAP2!: number;

  /** Surplus on Phase L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAS3!: number;

  /** Consumption in high tariff on Phase L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAH3!: number;

  /** Consumption in low tariff on Phase L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAL3!: number;

  /** Production on Phase L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAP3!: number;

  /** Surplus on Phase L1+L2+L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAS4!: number;

  /** Consumption in high tariff on Phase L1+L2+L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAH4!: number;

  /** Consumption in low tariff on Phase L1+L2+L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAL4!: number;

  /** Production on Phase L1+L2+L3 in kWh */
  @Transform(({ value }) => parseFloat(value))
  @IsDefined()
  @IsNumber()
  SAP4!: number;
}

export class AllTimeStatsShape {
  @Type(() => AllTimeStats)
  @ValidateNested()
  @IsDefined()
  // eslint-disable-next-line camelcase
  stat_alltime!: AllTimeStats;
}
