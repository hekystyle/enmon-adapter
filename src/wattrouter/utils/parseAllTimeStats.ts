import { plainToInstance } from 'class-transformer';
import { validateOrReject } from 'class-validator';
import { AllTimeStats, AllTimeStatsShape } from '../types.js';

export const parseAllTimeStats = async (plain: unknown): Promise<AllTimeStats> => {
  const instance = plainToInstance(AllTimeStatsShape, plain);
  await validateOrReject(instance);
  return instance.stat_alltime;
};
