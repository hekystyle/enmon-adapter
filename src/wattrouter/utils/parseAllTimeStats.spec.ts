import { parseAllTimeStats } from './parseAllTimeStats.js';
import { AllTimeStats } from '../types.js';

it.each([
  [{}],
  [{ stat_alltime: {} }],
  [
    {
      stat_alltime: {
        SAD: 'invalid date',
        SAS1: 'invalid number (SAS1)',
        SAH1: 'invalid number (SAH1)',
        SAL1: 'invalid number (SAL1)',
        SAP1: 'invalid number (SAP1)',
        SAS2: 'invalid number (SAS2)',
        SAH2: 'invalid number (SAH2)',
        SAL2: 'invalid number (SAL2)',
        SAP2: 'invalid number (SAP2)',
        SAS3: 'invalid number (SAS3)',
        SAH3: 'invalid number (SAH3)',
        SAL3: 'invalid number (SAL3)',
        SAP3: 'invalid number (SAP3)',
        SAS4: 'invalid number (SAS4)',
        SAH4: 'invalid number (SAH4)',
        SAL4: 'invalid number (SAL4)',
        SAP4: 'invalid number (SAP4)',
      },
    },
  ],
])('should reject invalid', async plain => {
  const promise = parseAllTimeStats(plain);
  await expect(promise).rejects.toBeInstanceOf(Array);
  await expect(promise).rejects.toMatchSnapshot();
});

it.each([
  [
    {
      stat_alltime: {
        SAD: '2022-08-28',
        SAH1: '1.01',
        SAH2: '2.02',
        SAH3: '3.03',
        SAH4: '4.04',
        SAL1: '5.05',
        SAL2: '6.06',
        SAL3: '7.07',
        SAL4: '8.08',
        SAP1: '9.09',
        SAP2: '10.10',
        SAP3: '11.11',
        SAP4: '12.12',
        SAS1: '13.13',
        SAS2: '14.14',
        SAS3: '15.15',
        SAS4: '16.16',
      },
    },
  ] as const,
])('should parse valid', async plain => {
  const promise = parseAllTimeStats(plain);
  await expect(promise).resolves.toBeInstanceOf(AllTimeStats);
  await expect(promise).resolves.toMatchSnapshot();
});
