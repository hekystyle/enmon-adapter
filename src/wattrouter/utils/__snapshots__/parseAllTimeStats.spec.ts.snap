// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse valid 1`] = `
{
  "SAD": "2022-08-28",
  "SAH1": 1.01,
  "SAH2": 2.02,
  "SAH3": 3.03,
  "SAH4": 4.04,
  "SAL1": 5.05,
  "SAL2": 6.06,
  "SAL3": 7.07,
  "SAL4": 8.08,
  "SAP1": 9.09,
  "SAP2": 10.1,
  "SAP3": 11.11,
  "SAP4": 12.12,
  "SAS1": 13.13,
  "SAS2": 14.14,
  "SAS3": 15.15,
  "SAS4": 16.16,
}
`;

exports[`should reject invalid 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Required",
    "path": [
      "stat_alltime",
    ],
    "received": "undefined",
  },
]
`;

exports[`should reject invalid 2`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Required",
    "path": [
      "stat_alltime",
      "SAD",
    ],
    "received": "undefined",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP4",
    ],
    "received": "nan",
  },
]
`;

exports[`should reject invalid 3`] = `
[
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP1",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP2",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP3",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAS4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAH4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAL4",
    ],
    "received": "nan",
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "message": "Expected number, received nan",
    "path": [
      "stat_alltime",
      "SAP4",
    ],
    "received": "nan",
  },
]
`;
